File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if((long)m*k>bloomDay.length){
            return -1;
        }
        int l=Integer.MAX_VALUE;
        int h=Integer.MIN_VALUE;
        for(int v:bloomDay){
            l=Math.min(l,v);
            h=Math.max(h,v);
        }
        int ans=h;
        while(l<=h){
            int mid=l+(h-l)/2;
            if(possible(mid,bloomDay,m,k)){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
    }
    public boolean possible(int mid,int[] bloomDay,int m ,int k){
        int count=0;
        int total=0;
        for(int i=0;i<bloomDay.length;i++){
            if(bloomDay[i]<=mid){
                count++;
            }
            else{
                count=0;
            }
            if(count==k){
                total++;
                count=0;
            }
            if(total>=m){
                return true;
            }
        }
        return false;
    }
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        if((long long)m*k>bloomDay.size()){
            return -1;
        }
        int l=INT_MAX;
        int h=INT_MIN;
        for(int x:bloomDay){
            l=min(l,x);
            h=max(h,x);
        }
        int ans=h;
        while(l<=h){
            int mid=l+(h-l)/2;
            if(poss(bloomDay,m,k,mid)){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
    }
    bool poss(vector<int>& arr,int m,int k,int day){
        int c=0;
        int bckt=0;
        for(int x:arr){
            if(x<=day){
                c++;
                if(c==k){
                    bckt++;
                    c=0;
                }
            }
            else{
                c=0;
            }
        }
        return bckt>=m;
    }
};